test/extended/util/metrics.go

```diff
@@ -0,0 +1,327 @@
+		"openshift-cluster-kube-scheduler-operator":               1.8,
+		"openshift-cluster-openshift-controller-manager-operator": 1.7,
+	}
+)
+
+type podInfo struct {
+	name      string
+	qps       float64
+	status    string
+	namespace string
+	result    string
+	failed    bool
+	skipped   bool
+}
+
+func CalculatePodMetrics(adminClient kubernetes.Interface, adminConfig *restclient.Config) error {
+	podUrlGetter := &PortForwardURLGetter{
+		Protocol:   "https",
+		Host:       "localhost",
+		RemotePort: "8443",
+		LocalPort:  "37587",
+	}
+
+	namespaces, err := adminClient.CoreV1().Namespaces().List(metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	failures := []error{}
+	for _, ns := range namespaces.Items {
+		// skip namespaces which do not meet "operator namespace" criteria
+		if !strings.HasPrefix(ns.Name, "openshift-") || !strings.HasSuffix(ns.Name, "-operator") {
+			continue
+		}
+
+		infos, err := getPodInfoForNamespace(adminClient, adminConfig, podUrlGetter, ns.Name)
+		if err != nil {
+			return err
+		}
+
+		for _, info := range infos {
+			if info.failed {
+				failures = append(failures, fmt.Errorf("Failed to fetch operator pod metrics for pod %q: %s", info.name, info.result))
+				continue
+			}
+			if info.skipped {
+				continue
+			}
+
+			qpsLimit := maxQPSAllowed
+			if customLimit, ok := perComponentNamespaceMaxQPSAllowed[info.namespace]; ok {
+				qpsLimit = customLimit
+			}
+
+			if info.qps > qpsLimit {
+				failures = append(failures, fmt.Errorf("operator pod %q in namespace %q is making %v requests per second. Maximum allowed is %v requests per second", info.name, info.namespace, info.qps, maxQPSAllowed))
+				continue
+			}
+		}
+	}
+
+	if len(failures) > 0 {
+		return errors.NewAggregate(failures)
+	}
+	return nil
+}
+
+func getPodInfoForNamespace(adminClient kubernetes.Interface, adminConfig *restclient.Config, podUrlGetter *PortForwardURLGetter, namespace string) ([]*podInfo, error) {
+	pods, err := adminClient.CoreV1().Pods(namespace).List(metav1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	podInfos := []*podInfo{}
+	for _, pod := range pods.Items {
+		info := &podInfo{
+			name:      pod.Name,
+			namespace: pod.Namespace,
+			status:    string(pod.Status.Phase),
+		}
+
+		podReady, err := utils.PodRunningReady(&pod)
+		if !podReady || err != nil {
+			result := "skipped, pod is not Running"
+			if err != nil {
+				result = fmt.Sprintf("%s: %v", result, err)
+			}
+
+			info.result = result
+			info.skipped = true
+			podInfos = append(podInfos, info)
+			continue
+		}
+
+		if len(pod.Spec.Containers) == 0 {
+			info.result = "skipped, no containers found"
+			info.skipped = true
+			podInfos = append(podInfos, info)
+			continue
+		}
+
+		metrics, err := getPodMetrics(adminConfig, &pod, podUrlGetter)
+		if err != nil {
+			// ignore errors from pods with no /metrics endpoint available
+			if !strings.Contains(err.Error(), "Connection refused") && !strings.Contains(err.Error(), "EOF") {
```

Comment:
please compare error types not string contents

Reply:
Changed this to a type switch where I match on *url.Error. Even then, however, I still have to do string matching to ensure the error is an EOF error. Sadly, I cannot do this more elegantly, as the error type is not the same as `io.EOF`, for example